### PR TITLE
add set body buffer for http client

### PR DESCRIPTION
**Pull request by liulanzheng**

### CODE REVIEW

File: net/http/client.h [hunk 1]
```diff
@@ -91,6 +94,12 @@ class Client : public Object {
             uds_path = unix_socket_path;
             return _client->call(this);
         }
+        // set body buffer and set content length automatically
+        void set_body(void *buf, size_t size) {
```
Comment (lihuiba) [inline net/http/client.h:98]: add an overloading for ```std::string_view```

File: net/http/client.h [hunk 2]
```diff
@@ -62,9 +62,12 @@ class Client : public Object {
         bool enable_proxy = false;
         std::string_view uds_path;                // If set, Unix Domain Socket will be used instead of TCP.
                                                   // URL should still be the format of http://localhost/xxx
-        IStream* body_stream = nullptr;                 // use body_stream as body
-        using BodyWriter = Delegate<ssize_t, Request*>; // or call body_writer if body_stream
-        BodyWriter body_writer = {};                    // is not set
+
+        void *body_buffer = nullptr;              // priority: body_buffer > body_stream > body_writer
```
Comment (lihuiba) [inline net/http/client.h:66]: body_buffer, body_buffer_size设为protected成员

Comment (lihuiba) [inline net/http/client.h:66]: set body_buffer, body_buffer_size as **protected** members

File: net/http/client.h [hunk 3]
```diff
@@ -106,6 +122,8 @@ class Client : public Object {
         explicit Operation(uint16_t buf_size) : req(_buf, buf_size), _client(nullptr) {}
         Operation() = delete;
         ~Operation() = default;
+
+        friend class ClientImpl;
```
Comment (lihuiba) [inline net/http/client.h:126]: ClientImpl is a sub-class of Client. There's no need be friend?

File: net/http/client.h [hunk 4]
```diff
@@ -91,9 +92,24 @@ class Client : public Object {
...
+        void set_body(std::string_view buf) {
+            body_buffer = (void*)buf.data();
+            body_buffer_size = buf.size();
+            req.headers.content_length(body_buffer_size);
+        }
+
 
     protected:
         Client* _client;
+        void *body_buffer = nullptr;
```
Comment (lihuiba) [inline net/http/client.h:110]: const void*

File: net/http/client.h [hunk 5]
```diff
@@ -91,9 +92,24 @@ class Client : public Object {
             uds_path = unix_socket_path;
             return _client->call(this);
         }
+        // set body buffer and set content length automatically
```
Comment (lihuiba) [inline net/http/client.h:95]: const void*

File: net/http/client.h [hunk 6]
```diff
@@ -91,9 +92,24 @@ class Client : public Object {
...
+        void set_body(void *buf, size_t size) {
+            body_buffer = buf;
+            body_buffer_size = size;
+            req.headers.content_length(size);
+        }
+        void set_body(std::string_view buf) {
+            body_buffer = (void*)buf.data();
```
Comment (lihuiba) [inline net/http/client.h:102]: return set_body(buf.data(), buf.length());

File: net/http/client.h (hunk 3 above)
Comment (liulanzheng) [reply net/http/client.h:126]: 编译不过啊

**Review by lihuiba — APPROVED** [review]

**Pull request merged**